### ISSUE TITLE
Which functions are concurrent-safe?

### ISSUE THREAD

**Issue by nowakf** [faiface/pixel#111]
It's not clear from the doccumentation which functions can be called in a goroutine, and which ones must be called in the main thread.

 I've been doing some experiments with stopping glfw - replacing glfw.PollEvents() with glfw.WaitEvents(), for instance - and some things appear to block (text.Draw(), for instance). 

I've tried grepping every instance of 'mainthread.Call', but it's still not clear.

**Comment by faiface** [comment]
Hi! Pixel handles all main-thread related stuff under the hood. You can use all Pixel functions and objects from as many goroutines as you like. There are a few rules, however:

1. A single Pixel object can't be used concurrently from multiple goroutines
2. Only a single operation runs on the main-thread at any given time

The second point implies that running a long operation on the main-thread will block it.

Pixel is made so that all functions sent, under the hood, to main-thread take very little time and so nothing blocks. If you, however, run `glfw.WaitEvents`, then that will block the main-thread, because `glfw.WaitEvents` waits (just as it says) until events arrive. In the meantime, nothing can be done.

**Comment by nowakf** [comment]
Thanks! That explains it - I was trying to use one text object from multiple goroutines. I had this idea of using glfw.PostEvent() so I could unblock the main thread from a goroutine whenever I want.

**Comment by faiface** [comment]
I mean, you *can* use one object from multiple goroutines, but it's a race condition, there's nothing protecting it.

**Comment by nowakf** [comment]
Yes - I was kinda proud of myself when I realized I'd created the dreaded race condition, and noticed it. It feels like I'm developing as a programmer - if probably in the wrong direction.

I actually tried just adding locks all over the place - but I think a more reasonable way to do it would be to not share objects between goroutines.

**Comment by faiface** [comment]
Okay, I guess we can close this?
